### DISCUSSION TITLE
Integration with Asp.net core self hosted

### DISCUSSION THREAD

**Discussion by ShadowCarlo** [vpnhood/VpnHood#93] [Q&A]
Hi,
I have two asp.net (.NET 5) projects both self hosted, which for convenience we will call them P1 and P2; P1 is installed on a server accessible from the internet via port 55371, while P2 is installed on a PC that connects to P1.
I would like to integrate VpnHood server in P1 and VpnHood Client in P2 so that P2 also acts as an http tunnel client and allows another service, present on the same PC where P2 is installed, to be visible to P1 which should make REST calls towards this service.

Do you think it is possible?

I hope my request is understandable,
Thank you
Carlo

**Comment by bobvhood** [comment]
Hi, I couldn't understand it clearly, let's forget about tunnels and VPN at the moment and focus on what you need to do first before finding a solution.

Do you just need to call some REST API on p1 which is not available to p2 via 55371?

**Comment by ShadowCarlo** [comment]
Hi Bob,
thank you for your interest, then the answer to your question is no and I try to describe the situation.
P1 and P2 communicate correctly on port 55371 and P2, even if it is another asp.net service, manages to make REST calls to P1 without problems.
On the PC where P2 is installed there is another REST server, which for convenience we will call O1, it is not owned by me but which I can query through the API I purchased.
The problem is that O1 needs to be called from P1 and I was thinking of using the HTTP TUNNELING technique to make O1 accessible from P1.
Of course you could ask me why I don't use P2 to query O1 and transfer data to P1, and I answer that for various technical problems, and not only, I cannot use this solution; I have to make sure that P1 can call O1's Rest API even though O1 is only visible in localhost.

Searching on the internet I found VpnHood which seems to me an extraordinary project (I take this opportunity to congratulate the developers) only that instead of using it as an external tool to my project, I would like to integrate it so as to have both my functions and the management of the TUNNELING to allow P1 to call O1.

Excuse me for being long-winded, but my situation is quite complex.

Thanks for your availability and patience.
Carlo

![image](https://user-images.githubusercontent.com/90213964/151526470-d582a9c6-ccbc-4367-ab63-9d9959ec8cba.png)

**Reply by bobvhood** [reply]
First of all, you don't need HTTP tunnel, the tunnel protocol between your servers is not important here, you can use any protocol.
VpnHood might work, but we actively exclude LocalHost requests, so you need to force O1 to run on a local IP, not localhost (127.0.0.01), but you said it is not possible.

Idea1: I believe you'd better look for a networking trick rather than using any kind of VPN. Such as mapping local IP to localhost and so on.
Idea2: Maybe you can find a proxy or REST API proxy that does it for you.
Idea3: You may achieve that by VpnHood, but you need to remove loopback from WinDivert packet filter and restrict it to O1 port:
https://github.com/vpnhood/VpnHood/blob/main/VpnHood.Client.Device.WinDivert/WinDivertPacketCapture.cs
![image](https://user-images.githubusercontent.com/92276967/151536081-daf14462-f888-4224-96cd-c67018d1f8bc.png)
By the way, redirecting all localhost requests to another machine may cause the failure of many of your local machine services. I really don't recommend you to look for that.

Unfortunately, we think this is a very special case, and working on it does not add useful functionality to VpnHood unless many users request it, which I really don't think so.

I hope you overcome that.

**Comment by ShadowCarlo** [comment]
Hi Bob,
ok I understand, if it exists I will try another solution . :sweat_smile:

Thank you very much and have a nice day :+1: :smiley:
Carlo
